google/cloud/storage/internal/async/reader_connection_factory.cc

```diff
@@ -0,0 +1,40 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/storage/internal/async/reader_connection_factory.h"
+
+namespace google {
+namespace cloud {
+namespace storage_internal {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+
+void UpdateGeneration(google::storage::v2::ReadObjectRequest& request,
+                      storage::Generation generation) {
+  if (request.generation() != 0 || !generation.has_value()) return;
+  request.set_generation(generation.value_or(0));
+}
+
+void UpdateReadRange(google::storage::v2::ReadObjectRequest& request,
+                     std::int64_t received_bytes) {
+  if (received_bytes == 0) return;
+  if (request.read_limit() != 0) {
+    request.set_read_limit(request.read_limit() - received_bytes);
+  }
```

Comment:
It is a little weird that a read request that reaches its limit turns into a read request without a limit.  Perhaps if the limit decrements to zero we should give it a negative value so that a re-use (even if it "cannot happen") would produce an error?

Reply:
Fixed I think.  I guess we could return a `Status` to indicate the bad situation, but it seems too unlikely to be a problem in practice.  In any case, with these changes it should not go undetected (though there may be more efficient ways to detect this "never happens" condition).